Copy-less RocksDBStore.ForkBlockIndexes()

longfin opened the pull request:
This PR address https://github.com/planetarium/libplanet/issues/1280. 

Basic idea is, re-use the previous chain instead of eliminating the full index copy when `RocksDBStore.ForkBlockIndexes()`. to accomplish this, it also changes `IterateIndexes()` and `IndexBlockHash()` to address the previous chain if needed.

~~Furthermore, all chains can be referenced by other chains, so `RocksDBStore.DeleteChainId()` needs to be modified as well. IMO, it can be possible if we have the reference for the source chain side too. but it may be too complicated to maintain... cc @planetarium/libplanet~~

Updated: I modified `RocksDBStore.DeleteChainId()` and related APIs (`.IterateIndexes()` and `.IndexBlockHash()`) to hide from public API when a chain with the fork is deleted.

These changes are compatible with the previous storage layout. so I wrote changelogs as "Behavioural Changes".

dahlia reviewed (commented):
What if a new fork is made from the existing fork made from another?  Does it trace back to the previous chain of the previous chain too?  We should have a test for such case.

longfin commented on Libplanet.RocksDBStore/RocksDBStore.cs:
```diff
@@ -356,44 +390,24 @@ BlockHash branchpoint
             {
-                _chainDb.Write(writeBatch);
-                writeBatch.Dispose();
+                _chainDb.Remove(k.Key(), destCf);
```
We need to handle the genesis block in `destinationChainId` already have if we don't want to remove it.

longfin commented:
> What if a new fork is made from the existing fork made from another? Does it trace back to the previous chain of the previous chain too? We should have a test for such case.

Yes, I intended to trace previous recursively. I'll add a test case for this case.

longfin commented:
@planetarium/libplanet I cleaned up codes. PTAL.

I should apply the same patch for `MonoRocksDBStore` after the first review.  (so it's still draft 😓 )

longfin commented:
> > What if a new fork is made from the existing fork made from another? Does it trace back to the previous chain of the previous chain too? We should have a test for such case.
> 
> Yes, I intended to trace previous recursively. I'll add a test case for this case.

On second thought, `MonoRocksDBStore` will work fine after this patch and IMO we should remove it after other projects relying on it ready... so it's okay to merge without fixes for `MonoRocksDBStore` now.

longfin commented:
/rebase

limebell commented:
There is chain B, which is forked from chain A, what happens if we delete chain A? Can we still iterate block indexes in chain B?

dahlia reviewed (commented):
Could you write a test case for the following scenario?

1. The branch A has several blocks: g → b1 → b2 → b3.
2. The branch B is forked from A: g → b1 → b2 → b3.
3. The branch B has more blocks after A's tip: g → b1 → b2 → b3 → b4 → b5.
4. The branch C has other several blocks than A (except for the genesis block): g → b1ʹ → b2ʹ → b3ʹ.
5. The branch A is swapped with the branch C:
    - A: g → b1ʹ → b2ʹ → b3ʹ
    - C: g → b1 → b2 → b3
6. Check if the branch B is not affected and still has the same blocks: g → b1 → b2 → b3 → b4 → b5.

longfin commented:
> There is chain B, which is forked from chain A, what happens if we delete chain A? Can we still iterate block indexes in chain B?

Yes.

limebell commented:
/rebase

longfin commented:
> Could you write a test case for the following scenario?
> 
> 1. The branch A has several blocks: g → b1 → b2 → b3.
> 2. The branch B is forked from A: g → b1 → b2 → b3.
> 3. The branch B has more blocks after A's tip: g → b1 → b2 → b3 → b4 → b5.
> 4. The branch C has other several blocks than A (except for the genesis block): g → b1ʹ → b2ʹ → b3ʹ.
> 5. The branch A is swapped with the branch C:
>    
>    * A: g → b1ʹ → b2ʹ → b3ʹ
>    * C: g → b1 → b2 → b3
> 6. Check if the branch B is not affected and still has the same blocks: g → b1 → b2 → b3 → b4 → b5.

- Swapping the chain isn't a store-level operation. it's defined as `BlockChain<T>.Swap()` method. 
  - on store API's perspective, `BlockChain<T>.Swap()` is just switching `BlockChain.Id` to newer and deleting previous by `Store.DeleteChainId`. 
- In the above scenario, A is not immediately swapped to C at store level, and A will be deleted by `BlockChain<T>.Swap()`. However, this also has a count forked to B, so it is not actually erased.

dahlia commented:
> - Swapping the chain isn't a store-level operation. it's defined as `BlockChain<T>.Swap()` method. 
> 
>   - on store API's perspective, `BlockChain<T>.Swap()` is just switching `BlockChain.Id` to newer and deleting previous by `Store.DeleteChainId`. 
> 
> - In the above scenario, A is not immediately swapped to C at store level, and A will be deleted by `BlockChain<T>.Swap()`. 
> 
>   - however, A has forked count to B, so it is not actually erased.


Fair enough. 👌🏼

longfin commented:
/rebase

dahlia commented:
/rebase

dahlia reviewed (approved):
It would be great if we merge this after checking if this work well with the existing 9c-main network.

longfin commented:
I've made some changes as below after testing on 9c mainnet 😅 

- Fork counting can be fragile against parallel execution. so I replaced `GetForks()` with `HasFork()`, `AddFork()` and `RemoveFork()` that require exact chain ID instead of scalar value.
- It has often been found that after a fork, no blocks are added and the block is forked again. To address this, if the index of the branch point is the same as the previous fork index, try to fork directly from the previous chain, not the current chain.
- Fixed a bug where chains deleted(marked) by `.DeleteChainId()` were output by `.ListChainIds()`.

I'll continue testing, but feel free to ask for your feedback. @libplanet

longfin commented:
I'm working on [the next version of this PR](https://github.com/longfin/libplanet.net/tree/feature/improve-fork-index-cf-removal). but it has breaking change for index layout so also requires data migration. 

Therefore, I'll merge this PR and open a new PR to continue. thanks for reviewing @planetarium/libplanet .
